processor/src/it/java/com/linecorp/decaton/processor/RetryQueueingIntegration.java

```diff
@@ -0,0 +1,96 @@
+
+import java.time.Duration;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.CountDownLatch;
+
+import org.junit.ClassRule;
+import org.junit.Test;
+
+import com.linecorp.decaton.client.DecatonClient;
+import com.linecorp.decaton.processor.runtime.ProcessorSubscription;
+import com.linecorp.decaton.processor.runtime.RetryConfig;
+import com.linecorp.decaton.protobuf.ProtocolBuffersDeserializer;
+import com.linecorp.decaton.protocol.Sample.HelloTask;
+import com.linecorp.decaton.testing.KafkaClusterRule;
+
+public class RetryQueueingIntegration {
+    @ClassRule
+    public static KafkaClusterRule rule = new KafkaClusterRule();
+
+    @Test(timeout = 30000)
+    public void testRetryQueuing() throws Exception {
+        String topicName = "test-" + UUID.randomUUID();
+        rule.admin().createTopic(topicName, 3, 3);
+        rule.admin().createTopic(topicName + "-retry", 3, 3);
+
+        Set<String> keys = new HashSet<>();
+        Set<String> keysNeedRetry = new HashSet<>();
+
+        // scenario:
+        //   * produce 100 tasks at total.
+        //   * half tasks have "-needRetry" suffix. they will be retried once, and processed after retried
+        //   * other tasks will be processed as usual
+        for (int i = 0; i < 50; i++) {
+            keysNeedRetry.add("key" + i + "-needRetry");
+        }
+        for (int i = 50; i < 100; i++) {
+            keys.add("key" + i);
+        }
+        Set<String> processedKeys = Collections.synchronizedSet(new HashSet<>());
+        Set<String> retriedKeys = Collections.synchronizedSet(new HashSet<>());
+        CountDownLatch processLatch = new CountDownLatch(keys.size() + keysNeedRetry.size());
+
+        DecatonProcessor<HelloTask> processor = (context, task) -> {
+            String key = context.key();
+            if (!key.endsWith("needRetry")) {
+                processedKeys.add(key);
+                processLatch.countDown();
+                return;
+            }
+            if (context.metadata().retryCount() == 0) {
+                context.retry();
+            } else {
+                retriedKeys.add(key);
+                processLatch.countDown();
+            }
+        };
+
+        try (ProcessorSubscription subscription = TestUtils.subscription(
+                rule.bootstrapServers(),
+                ProcessorsBuilder.consuming(topicName, new ProtocolBuffersDeserializer<>(HelloTask.parser()))
+                                 .thenProcess(processor),
+                RetryConfig.withBackoff(Duration.ofMillis(10)));
+             DecatonClient<HelloTask> client = TestUtils.client(topicName, rule.bootstrapServers())) {
+
+            keys.forEach(key -> client.put(key, HelloTask.getDefaultInstance()));
+            keysNeedRetry.forEach(key -> client.put(key, HelloTask.getDefaultInstance()));
+
+            processLatch.await();
+
+            assertEquals(keys, processedKeys);
```

Comment:
no point to put those check in this try block?

Reply:
I'm guessing if it's good idea to test equity of two sets which contains 50 elements.
Let's say that this test fails, then it outputs the result of comparison for 50 elements, which is already difficult to glance by eyes but even if we could noticed that the particular key is missing here, would it be helps us to figure out the cause?

Having this test to fail, I imagine we'll anyway have to take some effort to enhance debugging information to and reproduce the issue several times so here it would be enough to compare just the size of two sets? 

Besides, I think we want to increase test cases to have much more than 50, as in one possible situation that we want to catch in integration test is race conditions, that makes more sense to have many items to test.

Reply:
that's right. nice catch
> no point to put those check in this try block?

hm, that's completely enough
> compare just the size of two sets?